src/bt/Blackboard.cpp

```diff
@@ -12,9 +12,14 @@ void Blackboard::setBool(std::string key, bool value) {
 }
 
 bool Blackboard::getBool(std::string key) {
-    if (bools.find(key) == bools.end()) {
-        bools[key] = false;
-    }
+    // if (bools.find(key) == bools.end()) {
+    //     bools[key] = false;
+    // }
+    // bools[key] deafult constructs if no exist
+    // https://en.cppreference.com/w/cpp/container/map/operator_at
+    // If an insertion is performed, the mapped value is value-initialized (default-constructed for class types, zero-initialized otherwise) and a reference to it is returned.
+
+    // https://godbolt.org/z/87GUTV
```

Comment:
I'm pretty sure in this class it's actually intended behavior to return 0/false/empty string if the value is not found. This is one of our most ugly classes though, so I don't blame you.

Reply:
I know, and that's exactly what happens if the value is not found

bools[key] is explicitly set to false however operator[] will already do that for you.

Reply:
I'm not sure if I like that personally, as it makes readability more difficult. If you change it I think it's at least worth a comment since I don't expect the average programmer in our system to know that the access operator also zero-initializes a field.

Reply:
Okay, well the issue is not just that it makes code look more messy / better.

std::map::find() on itself is already O(n log n), then operator[] is another O(n log n) complexity, yes the complexity stays the same but the actual time it takes to perform 2 O(n log n) operations would double, it's not merely a readability thing. But yeah, it definitely could use a comment since looking at it for the first time without any operator[] knowledge could cause some confusion.

Reply:
I fully agree we should have your proposed changes here with a comment. Readability vs efficiency is an important discussion to have generally. The reason why you see so little unordered maps, constexpr and stuff like that is because most computations and skills do not really take up significant amounts of computing time in our system. 90+% of the computational time costs comes from 4 key algorithms. 

We however spend a lot more time fixing bugs and mistakes caused by difficult to read code and just bad code practices than optimizing our code, because in most places optimization is not really necessary. Only the code that is consistently run very often (these 4 algorithms) needs to be fast and efficient. In places like this, where the map is often very small, the difference between  O log(1) vs O log(n) vs O(n) or 2 O(n log(n)) is minimal, so we typically do not spend that much time worrying about performance, and as a consequence this code is often not very efficient. 

This is merely an explanation (and not a rebuttal or counterargument, I fully agree!), but keep it in mind if you want to go through more code, so you do not get frustrated. For us, human mistakes are a far bigger issue than program performance.